rust/chains/hyperlane-ethereum/src/rpc_clients/fallback.rs

```diff
@@ -119,15 +265,18 @@ impl JsonRpcClient for FallbackProvider<PrometheusJsonRpcClient<Http>> {
             if !errors.is_empty() {
                 sleep(Duration::from_millis(100)).await;
             }
-            for (idx, provider) in self.0.iter().enumerate() {
+            let priorities_snapshot = take_priorities_snapshot(self).await;
+            for (idx, priority) in priorities_snapshot.iter().enumerate() {
+                let provider = &self.inner.providers[priority.index];
                 let fut = match params {
                     Value::Null => provider.request(method, ()),
                     _ => provider.request(method, &params),
                 };
-
                 let resp = fut.await;
+                handle_stalled_provider(self, priority).await?;
```

Comment:
Wdyt of doing the stall check before making the actual RPC request, and if the provider in question is deprioritized we fallback to the next provider? Iiuc here we still treat the response from a possibly stalled provider as valid, and only get the "gains" of falling back to another provider in a subsequent RPC. Not a big deal but feels like we can get more for free by a small change

Reply:
Spent some time trying to add this but it's taking longer than it's worth it. The biggest problem is that the mocked provider's `request` will never get to sleep so that the deprioritization condition is met. The other issue is that it's taken me a long time to find a type that's `DeserializeOwned` so as to mock a return value that represents a real block and not just an error. 

I also realised that unless I can create a value that implements `DeserializedOwned`, I won't be able to test all the edge cases, because the providers inside `FallbackProvider` are all initialized with `Instant::now()` and tested serially, so if one sleeps for longer than the max block period, all of them will have become potentially stalled - and then `request` will return an error from `serde_json` returning an error in my mock.